docs/proposals/20210720-kdm-template.md

```diff
@@ -0,0 +1,109 @@
+
+1. The user applies the `KubeadmControlPlane` CR and all other necessary CAPI CRs.
+2. The `KubeadmControlPlane` controller keeps checking if the `NestedCluster` CR is ready.
+3. If yes, the `KubeadmControlPlane` controller use the `NestedMachineTemplate` to creates desired number of `NestedMachines` for the `NestedCluster` Infrastructure.
+4. The `KubeadmControlPlane` controller generates the `KubeadmConfig` CR based on the `KubeadmControlPlane` created in the first step. 
+5. The `KubeadmControlPlane` controller assigns the `KubeadmConfig` to the owned `NestedMachine` CR by setting the ObjectReference.
+6. The `NestedMachine` controller keeps checking if the `KubeadmConfig` has been assigned to the `NestedMachine` CR.
+7. If yes, the `NestedMachine` controller executes the `kubeadm` executable to generate the static pod template.
+8. The `NestedMachine` controller decorates the static pod template with the proper volume mounts and creates the raw pod objects, which allows the `KubeadmControlPlane` to manage these pods.
+
+![Templating the control plane](images/nestedcontrolplane/kdm-template.png)
+
+### User Stories
+
+1. As a user, I would like to have all my control plane components using standardized KubeadmControlPlane spec.
+1. As a user, I would like to customize the command line arguments of my `NestedControlPlane`'s components through a `KubeadmControlPlane` CR.
+2. As a user, I would like to customize the pod volumes of my `NestedControlPlane`'s components through a `KubeadmControlPlane` CR.
+
+### Implementation Details
+
+The provisioning of the control plane requires the involvement of the `NestedMachine` controller. We will discuss the `NestedMachine` controller and the complete provisioning process in another CAEP. 
+This CAEP will focus on how to template the control plane in a more "CAPI" way, i.e., using `KubeadmControlPlane` provider and `kubeadm` libs from the `kubernetes` module.  
+
+- Bake the `kubeadm` executable into the container image of the `NestedControlPlane` controller (we will import the functions instead of embedding the binary, once `kubeadm` is independent from the `Kubernetes`).
+- Use the existing `NestedControlPlane` implementation for testing, which uses the NestedKASController, NestedKCMController and the NestedEtcdController.
+- Create `KubeadmConfig` CR manually.
+- Implement the generic functionality of extracting the`kubeadmapi.ClusterConfig` from the `KubeadmConfig` CR.
+- Generate a config file that will be used by the `kubeadm` to generate the manifests.
+- Before adopting any nested component, `NestedControlPlane` will run the `kubeadm init --config` command to generate the static pod template for each component and create a configmap to store them.
```

Comment:
i think the exact cmd / `init` phase that is needed here is:
`kubeadm init phase control-plane --config`